zigzag/opt/loma/multipermute.py

```diff
@@ -51,6 +54,46 @@ def nth(self, n: int):
         return o
 
 
+class PermutationConstraint(ABC):
+    """! An abstract class to represent a constraint on a permutation."""
+
+    def is_valid(self, permutation: list[Any]) -> bool:
+        """! Returns True if the permutation satisfies the constraint."""
+        raise NotImplementedError
+
+
+class StaticPositionsConstraint(PermutationConstraint):
+    """! A class to represent a constraint on a permutation that requires
+    a predefined order for some or all elements."""
+
+    static_positions: dict[int, LayerDim]
+
+    def __init__(self, static_positions: dict[int, LayerDim]):
+        self.static_positions = static_positions
+
+    def is_valid(self, permutation: list[Any]) -> bool:
+        for position, item in self.static_positions.items():
+            if permutation[position][0] != item:
+                return False
+        return True
+
+
+class StaticPositionsAndSizesConstraint(PermutationConstraint):
+    """! A class to represent a constraint on a permutation
+    that requires a predefined order and size for some or all elements."""
+
+    static_positions_and_sizes: dict[int, tuple[LayerDim, int]]
+
+    def __init__(self, static_positions_and_sizes: dict[int, tuple[LayerDim, int]]):
+        self.static_positions_and_sizes = static_positions_and_sizes
+
+    def is_valid(self, permutation: list[Any]) -> bool:
+        for position, (item, size) in self.static_positions_and_sizes.items():
+            if permutation[position][0] != item or permutation[position][1] != size:
+                return False
+        return True
```

Comment:
```suggestion
    return all(permutation[position][0] == item for position, item in self.static_positions.items())
```

Should be a bit faster :)